### PR TITLE
Intial transfert of code

### PR DESCRIPTION
**Pull request by frheault**
To leave a trace, basic comments on architecture required

Here is some data, launching command and a singularity
_/braindata/processedData/frheault/flow_PR.zip_

### CODE REVIEW

File: USAGE [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+--root=/path/to/[root]              Root folder containing multiple subjects
+
+                                    [root]
+                                    ├── S1
+                                    │   ├── *anat.nii.gz
+                                    │   └── *tracking.trk
+                                    └── S2
+                                        └── *
+
+--atlas_config = Path to json config file
```
Comment (arnaudbore) [inline USAGE]: These options need to be more explicit. I've never used rbx and I'm lost.

File: main.nf [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+log.info "Input: $params.root"
+root = file(params.root)
+/* Watch out, files are ordered alphabetically in channel */
+    in_data = Channel
+        .fromFilePairs("$root/**/{*tracking.trk,*anat.nii.gz}",
+                        size: 2,
+                        maxDepth:2,
+                        flat: true) {it.parent.name}
+
+    atlas_centroids = Channel.fromPath("$params.atlas_centroids/*.trk")
```
Comment (arnaudbore) [inline main.nf]: maybe: `if atlas_centroids`

File: main.nf [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    input:
+    set sid, file(native_anat), file(atlas) from anats_for_registration
+
+    output:
+    set sid, "${sid}__output0GenericAffine.txt" into transformation_for_recognition, transformation_for_centroids
+    file "${sid}__outputWarped.nii.gz"
+    script:
+    """
+    antsRegistrationSyNQuick.sh -d 3 -f ${native_anat} -m ${atlas} -n ${params.register_processes} -o ${sid}__output -t a
+    ConvertTransformFile 3 ${sid}__output0GenericAffine.mat ${sid}__output0GenericAffine.txt --hm --ras
```
Comment (arnaudbore) [inline main.nf:103]: won't be usefull next week

**Review by arnaudbore — COMMENTED** [review]
First few comments, it would be very useful to have a small dataset to play with and have more documentations. You can put this on braindata if you want.

File: USAGE [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+--atlas_anat = Path to reference anatomy of the atlas
+--atlas_directory = Path of the folder containing all atlases
+
+NOTES
+
+OPTIONAL ARGUMENTS (current value)
+--atlas_centroids = Folder containing centroids (trk) for future tractometry (Must have the same filename as in the atlas_config)
```
Comment (GuillaumeTh) [inline USAGE]: Please format the documentation as below.

File: USAGE [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+--multi_parameters                      Number of executions for the multi-parameters ($multi_parameters)
+--minimal_vote_ratio                    Percentage of the vote to obtain in order to be considered valid ($minimal_vote_ratio)
+--wb_clustering_thr                     List of distance (mm) for QBx whole brain clustreing ($wb_clustering_thr)
+--seeds                                 List of random seeds values for the rng ($seeds)
+--outlier_alpha                         Remove spurious streamlines based on shape ($outlier_alpha)
```
Comment (GuillaumeTh) [inline USAGE:32]: Maybe add the number of threads options here too.

File: main.nf [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    .separate(3)
+
+workflow.onComplete {
+    log.info "Pipeline completed at: $workflow.complete"
+    log.info "Execution status: ${ workflow.success ? 'OK' : 'failed' }"
+    log.info "Execution duration: $workflow.duration"
+}
+
+anat_for_registration
+    .combine(atlas_anat)
```
Comment (GuillaumeTh) [inline main.nf:90]: Try with concat instead of combine. It is more clear.

File: main.nf [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    script:
+    """
+    antsRegistrationSyNQuick.sh -d 3 -f ${native_anat} -m ${atlas} -n ${params.register_processes} -o ${sid}__output -t a
+    ConvertTransformFile 3 ${sid}__output0GenericAffine.mat ${sid}__output0GenericAffine.txt --hm --ras
+    """ 
+}
+
+
+anat_for_reference
+    .combine(transformation_for_centroids, by: 0)
```
Comment (GuillaumeTh) [inline main.nf]: Why combine and not join ?

File: main.nf [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    output:
+    file "${sid}__${centroid.baseName}.trk"
+    script:
+    """
+    scil_apply_transform_to_tractogram.py ${centroid} ${anat} ${transfo} ${sid}__${centroid.baseName}.trk --inverse --remove_invalid
+    """ 
+}
+
+tractogram_for_recognition
+    .combine(transformation_for_recognition, by: 0)
```
Comment (GuillaumeTh) [inline main.nf:124]: Same here. Why not join ?

File: main.nf [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    .combine(atlas_config)
+    .combine(atlas_directory)
```
Comment (GuillaumeTh) [inline main.nf:126]: Please use concat for this case.

File: main.nf [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    mkdir tmp/
+    scil_recognize_multi_bundles.py ${tractogram} ${config} ${directory}/*/ ${transfo} --inverse --output tmp/ \
+        --log_level DEBUG --multi_parameters $params.multi_parameters --minimal_vote_ratio $params.minimal_vote_ratio \
+        --tractogram_clustering_thr $params.wb_clustering_thr --seeds $params.seeds --processes $params.rbx_processes
+    mv tmp/* ./
+    """ 
+}
+
+bundles_for_cleaning
+       .transpose()
```
Comment (GuillaumeTh) [inline main.nf]: Remove one tab.

File: main.nf [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    input:
+    set sid, file(bundle) from all_bundles_for_cleaning
+    output:
+    file "${sid}__*_cleaned.trk"
+    script:
+    bname = bundle.name.take(bundle.name.lastIndexOf('.'))
+    """
+    scil_outlier_rejection.py ${bundle} "${sid}__${bname}_cleaned.trk" --alpha $params.outlier_alpha
+    """ 
+}
```
Comment (GuillaumeTh) [inline main.nf]: Add new line.

File: main.nf [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    antsRegistrationSyNQuick.sh -d 3 -f ${native_anat} -m ${atlas} -n ${params.register_processes} -o ${sid}__output -t a
+    ConvertTransformFile 3 ${sid}__output0GenericAffine.mat ${sid}__output0GenericAffine.txt --hm --ras
+    """ 
+}
+
+
+anat_for_reference
+    .combine(transformation_for_centroids, by: 0)
+    .set{anat_and_transformation}
+process Transform_Centroids {
```
Comment (GuillaumeTh) [inline main.nf:111]: Add number of cpus to be safe.

File: nextflow.config [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    // Recobundle segmentation
+    multi_parameters = 18
+    minimal_vote_ratio = 0.5
+    wb_clustering_thr = "15 12"
+    seeds = "0"
+    outlier_alpha = 0.5
+
+    register_processes=4
+    rbx_processes=4
+}
```
Comment (GuillaumeTh) [inline nextflow.config]: Add new line.

**Review by GuillaumeTh — CHANGES_REQUESTED** [review]
Not tested yet but looks good !

File: USAGE (hunk 1 above)
Comment (frheault) [reply USAGE]: Should I show an example in the USAGE ?
Should I provide a fake example in the repo (its like 5ko anyway)

File: main.nf (hunk 2 above)
Comment (frheault) [reply main.nf]: I have to define atlas_centroids to an empty channel anyway.
So if would be 
```
if $params.atlas_centroids
   atlas_centroids = Channel.fromPath(...)
else
   atlas_centroids = Channel.empty()
```
Is that what you both want ?

File: main.nf (hunk 3 above)
Comment (frheault) [reply main.nf:103]: Then, merge it ;)

**Pull request merged**